app/models/gateway-account-credential/GatewayAccountCredential.class.js

```diff
@@ -9,15 +9,63 @@ const CREDENTIAL_STATE = {
 }
 
 class GatewayAccountCredential {
-  constructor (data) {
-    this.externalId = data.external_id
-    this.paymentProvider = data.payment_provider
-    this.credentials = Credential.fromJson(data.credentials)
-    this.state = data.state
-    this.createdDate = data.created_date
-    this.activeStartDate = data.active_start_date
-    this.activeEndDate = data.active_end_date
-    this.gatewayAccountId = data.gateway_account_id
+  withExternalId (externalId) {
+    this.externalId = externalId
+    return this
+  }
+
+  withPaymentProvider (paymentProvider) {
+    this.paymentProvider = paymentProvider
+    return this
+  }
+
+  withCredentials (credentials) {
+    this.credentials = credentials
+    return this
+  }
+
+  withState (state) {
+    if (state) {
+      this.state = state
+    }
+    return this
+  }
+
+  withCreatedDate (createdDate) {
+    this.createdDate = createdDate
+    return this
+  }
+
+  withActiveStartDate (activeStartDate) {
+    this.activeStartDate = activeStartDate
+    return this
+  }
+
+  withActiveEndDate (activeEndDate) {
+    this.activeEndDate = activeEndDate
+    return this
+  }
+
+  withGatewayAccountId (gatewayAccountId) {
+    this.gatewayAccountId = gatewayAccountId
+    return this
+  }
+
+  /**
+   *
+   * @param data
+   * @returns {GatewayAccountCredential}
+   */
+  static fromJson (data) {
+    return new GatewayAccountCredential()
+      .withExternalId(data?.external_id)
+      .withPaymentProvider(data?.payment_provider)
+      .withCredentials(Credential.fromJson(data?.credentials))
```

Comment:
how's this work for cases where an account has more than one credential?

Reply:
Those are separate `GatewayAccountCredential` classes, see https://github.com/alphagov/pay-selfservice/blob/54b4cc89e0a580ca82c4183fb0e62e3b40a36b49/app/models/GatewayAccount.class.js#L45-L47